sql/logictest: TestLogic failed

[(sql/logictest).TestLogic failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2270164&tab=buildLog) on [master@5ec2eab07254c4bbd05902a77d73809cbf1f9868](https://github.com/cockroachdb/cockroach/commits/5ec2eab07254c4bbd05902a77d73809cbf1f9868):

```
=== RUN   TestLogic
--- FAIL: TestLogic (775.39s)
    logic.go:2812: randomize batchSize to 3
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLogic999722559
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestLogic/local-vec-auto/sequences
=== PAUSE TestLogic/local-vec-auto/sequences
=== CONT  TestLogic/local-vec-auto/sequences
        --- FAIL: TestLogic/local-vec-auto/sequences (11.61s)
=== RUN   TestLogic/local-vec-auto
    --- FAIL: TestLogic/local-vec-auto (0.39s)
=== RUN   TestLogic/local-vec-auto/sequences/generator_timeout
--- done: testdata/logic_test/sequences with config local-vec-auto: 208 tests, 2 failures
            --- FAIL: TestLogic/local-vec-auto/sequences/generator_timeout (0.03s)
                logic.go:2347: 
                    
                    testdata/logic_test/sequences:900: 
                    expected success, but found
                    (57014) query execution canceled due to statement timeout
                    errors.go:213: in init()
                logic.go:1660: 
                     pq: query execution canceled due to statement timeout

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestLogic PKG=./pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestLogic.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/conn_executor_exec.go
+++ b/pkg/sql/conn_executor_exec.go
@@ -400,7 +400,9 @@ func (ex *connExecutor) execStmtInOpenState(
 		return ev, payload, nil
 	}
 
-	if ex.sessionData.StmtTimeout > 0 {
+	// We exempt `SET` statements from the statement timeout, particularly so as
+	// not to block the `SET statement_timeout` command itself.
+	if ex.sessionData.StmtTimeout > 0 && stmt.AST.StatementTag() != "SET" {
 		timerDuration := ex.sessionData.StmtTimeout - timeutil.Since(ex.phaseTimes[sessionQueryReceived])
 		// There's no need to proceed with execution if the timer has already expired.
 		if timerDuration < 0 {

--- a/pkg/sql/logictest/testdata/logic_test/set
+++ b/pkg/sql/logictest/testdata/logic_test/set
@@ -290,6 +290,25 @@ SHOW statement_timeout
 ----
 100
 
+# Set the statement timeout to something absurdly small, so that no query would
+# presumably be able to go through. It should still be possible to get out of
+# this "bad state" by resetting the statement timeout.
+subtest impossible_statement_timeout_recovery
+
+statement ok
+SET statement_timeout = '1us'
+
+statement error query execution canceled due to statement timeout
+SHOW statement_timeout
+
+statement ok
+SET statement_timeout = 0
+
+query T
+SHOW statement_timeout
+----
+0
+
 # Test that composite variable names get rejected properly, especially
 # when "tracing" is used as prefix.
 

